include/realizations/catchment/Formulation_Manager.hpp

```diff
@@ -285,17 +291,28 @@ namespace realization {
              * 
              * @return std::string of the output root directory
              */
-            std::string get_output_root() const noexcept {
+            std::string get_output_root() const {
                 const auto output_root = this->tree.get_optional<std::string>("output_root");
+                std::string str;
```

Comment:
I think this should fix your issue.
```suggestion
                std::string str = "./";
```

Reply:
You mean the failing tests or something else. I don't want to write to "./" though.

Reply:
I'm fairly certain the call to `stat` will fail if you provide it the empty string like you are right now in the case when `output_root` is not present. I think that is why the `Build_and_Run_Model` action is failing.

I think you should return `"./"` if `output_root` is not present (`boost::none`) like in the previous code. We can always assume that the directory `ngen` is run from, `./` , exists. 

One solution is:

```c++
                if (output_root != boost::none && *output_root != "") {
                    // Check if the path ends with a trailing slash,
                    // otherwise add it.
                    output_root->back() == '/'
                           ? *output_root
                           : *output_root + "/";

                     //use C system function to check if there is a dir match that defined in realization
                    struct stat sb;
                    if (stat(output_root->c_str(), &sb) == 0 && S_ISDIR(sb.st_mode))
                        return *output_root;
                    else {
                        throw std::runtime_error("output_root directory does not exist, please create one matching that in realization");
                    }
                }
                return "./";
```

Aside, I would think we would want to check that the calculated `output_root` can be written to by the current user. That is out of the scope of this PR and should be handled elsewhere if deemed necessary.

Reply:
(1) It looks like this code would return "./" if output_root directory does not exist.
(2) This code would seem not get a chance to throw based on the outer if, which is not what we want.
(3) This block of code is a bit tricky:
```
                    output_root->back() == '/'
                           ? *output_root
                           : *output_root + "/";
```
where does the value land?
(4) The failed test looks like come from this line in Example_model_run.yml:
     ` ./ngen data/catchment_data.geojson '' data/nexus_data.geojson '' data/example_realization_config.json`
It takes realization file from the master, not from the PR, right?
(5) In the latest realization file, I set output_root to "./", so it automatically exists.

Reply:
This code handles the output root for `ngen`, so if there is one in a realization config file under the key, `output_root`, that one will be used, otherwise, the default is to write to the directory where `ngen` was called, `"./"`.

I don't think you pushed an updated `data/example_realization_config.json` file. Looks like you pushed up `data/example_bmi_multi_realization_config.json` and `data/example_bmi_multi_realization_config_w_noah_pet_cfe.json`.